packages/flutter_tools/lib/src/proxied_devices/devices.dart

```diff
@@ -288,9 +304,33 @@ class ProxiedDevice extends Device {
     final String fileName = binary.basename;
     final Completer<String> idCompleter = Completer<String>();
     _applicationPackageMap[path] = idCompleter.future;
-    await connection.sendRequest('proxy.writeTempFile', <String, Object>{
-      'path': fileName,
-    }, await binary.readAsBytes());
+
+    final Map<String, Object> args = <String, Object>{'path': fileName};
+
+    Map<String, Object?>? rollingHashResultJson;
+    if (_deltaFileTransfer) {
+     rollingHashResultJson = _cast<Map<String, Object?>?>(await connection.sendRequest('proxy.calculateFileHashes', args));
+    }
+
+    if (rollingHashResultJson == null) {
+      // Either file not found on the remote end, or deltaFileTransfer is set to false, transfer the file directly.
```

Comment:
should we put a debug log if `_deltaFileTransfer == true && rollingHashResultJson == null`?

Reply:
This is normal when it's the first run. But it's probably still a good idea to add a log, just added, thanks!

Reply:
oh good point about first run.